keycloak/lib/stacks/keycloak.ts

```diff
@@ -0,0 +1,164 @@
+/**
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+import { Duration, Stack, StackProps } from 'aws-cdk-lib';
+import {
+  AutoScalingGroup, BlockDeviceVolume, Monitoring, Signals,
+} from 'aws-cdk-lib/aws-autoscaling';
+import {
+  AmazonLinuxCpuType, CloudFormationInit, InitCommand, InitElement, InitFile, InitPackage,
+  InstanceClass, InstanceSize, InstanceType, MachineImage, Peer, Port, SecurityGroup, SubnetType, Vpc,
+} from 'aws-cdk-lib/aws-ec2';
+import {
+  ApplicationLoadBalancer, ApplicationProtocol, ListenerCertificate, Protocol, SslPolicy,
+} from 'aws-cdk-lib/aws-elasticloadbalancingv2';
+import {
+  ManagedPolicy, PolicyStatement, Role, ServicePrincipal,
+} from 'aws-cdk-lib/aws-iam';
+import { ARecord, RecordTarget } from 'aws-cdk-lib/aws-route53';
+import { Construct } from 'constructs';
+import { join } from 'path';
+import { LoadBalancerTarget } from 'aws-cdk-lib/aws-route53-targets';
+import { KeycloakUtils } from './utils';
+
+export interface ALBprops {
+  certificateArn: string;
+  hostedZone: KeycloakUtils;
+}
+export interface KeyCloakProps extends StackProps {
+  vpc: Vpc;
+  keycloakSecurityGroup: SecurityGroup;
+  rdsInstanceEndpoint: string;
+  keycloakDBpasswordSecretArn?: string;
+  keycloakAdminUserSecretArn: string;
+  keycloakAdminPasswordSecretArn: string;
+  keycloakCertPemSecretArn: string;
+  keycloakCertKeySecretArn: string;
+  albProps: ALBprops
+}
+
+export class KeycloakStack extends Stack {
+  readonly loadBalancerARN: string
+
+  constructor(scope: Construct, id: string, props: KeyCloakProps) {
+    super(scope, id, props);
+
+    const instanceRole = this.createInstanceRole();
+
+    const keycloakNodeAsg = new AutoScalingGroup(this, 'keycloakASG', {
+      instanceType: InstanceType.of(InstanceClass.C5, InstanceSize.XLARGE9),
```

Comment:
`c5.9xlarge` is pretty HUGE instance type? Are we really expecting such traffic? Can we start with `c5.2xlarge` and then decide whether we want to scale up or not?

Reply:
Since keycloak is using external database I kept some buffer with high CPU and memory

Reply:
Didn't get the external db part for adding buffer.

Reply:
Not sure how much memory does database operations need. I can reduce and then observe if we need more.

Reply:
Reduced to c5.2xlarge

Reply:
depends how much data is being queried. What kind of data will be stored and what sort of queries will be made will give an idea.

Reply:
All the data that keycloak has is written to external data base. However all of this is running in a container so I think we should be good with reduced size for now. Let's see how it goes.